File: src/test/java/com/alvcohen/travelhelper/service/UserServiceTest.java
package com.alvcohen.travelhelper.service;

import com.alvcohen.travelhelper.dto.VisaUsageAddRequest;
import com.alvcohen.travelhelper.model.AuthProvider;
import com.alvcohen.travelhelper.model.User;
import com.alvcohen.travelhelper.model.VisaUsage;
import com.alvcohen.travelhelper.repository.UserRepository;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.junit4.SpringRunner;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

@RunWith(SpringRunner.class)
@SpringBootTest
@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
public class UserServiceTest {

    @Autowired
    private UserService userService;

    @Autowired
    private UserRepository userRepository;

    @Before
    public void setUp() {
        final User user = new User("testuser@example.com", "TestUser", AuthProvider.LOCAL);
        userRepository.save(user);
    }

    @Test
    public void addVisaUsage_addsUsageWithDaysThatIncludeArrival() {
        userService.addVisaUsage("testuser@example.com", new VisaUsageAddRequest("01/02/2020", "07/02/2020"));
        final User updatedUser = userService.getCurrentUser("testuser@example.com");
        assertFalse(updatedUser.getVisaUsages().isEmpty());
        final VisaUsage usage = updatedUser.getVisaUsages().iterator().next();
        assertEquals(1, usage.getArrival().getDayOfMonth());
        assertEquals(2, usage.getArrival().getMonthValue());
        assertEquals(7, usage.getDeparture().getDayOfMonth());
        assertEquals(2, usage.getDeparture().getMonthValue());
        assertEquals(7, usage.getDays());
    }

    @Test
    public void deleteVisaUsageById_cascadeDeletesUsageFromUser() {
        userService.addVisaUsage("testuser@example.com", new VisaUsageAddRequest("01/02/2020", "07/02/2020"));
        final User updatedUser = userService.getCurrentUser("testuser@example.com");
        assertEquals(1, updatedUser.getVisaUsages().size());
        final VisaUsage usage = updatedUser.getVisaUsages().iterator().next();
        userService.deleteVisaUsageById(usage.getId());
        final User userWithDeletedUsage = userService.getCurrentUser("testuser@example.com");
        assertTrue(userWithDeletedUsage.getVisaUsages().isEmpty());
    }
}

File: src/test/kotlin/com/alvcohen/travelhelper/services/UserServiceTest.kt
package com.alvcohen.travelhelper.services

import com.alvcohen.travelhelper.AuthProvider
import com.alvcohen.travelhelper.User
import com.alvcohen.travelhelper.UserRepository
import com.alvcohen.travelhelper.VisaUsage
import com.alvcohen.travelhelper.controllers.VisaUsageAddRequest
import junit.framework.Assert.assertEquals
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.context.junit4.SpringRunner

@RunWith(SpringRunner::class)
@SpringBootTest
@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
class UserServiceTest {

    @Autowired
    private lateinit var userService: UserService

    @Autowired
    private lateinit var userRepository: UserRepository

    @Before
    fun setUp() {
        val user = User("testuser@example.com", "TestUser", AuthProvider.LOCAL)
        userRepository.save(user)
    }

    @Test
    fun addVisaUsage_addsUsageWithDaysThatIncludeArrival() {
        userService.addVisaUsage("testuser@example.com", VisaUsageAddRequest("01/02/2020", "07/02/2020"))
        val updatedUser: User? = userService.getCurrentUser("testuser@example.com")
        Assert.assertFalse(updatedUser?.let { updatedUser.visaUsages.isEmpty() } ?: false)
        val usage: VisaUsage? = updatedUser?.let{updatedUser.visaUsages.first()}
        assertEquals(1, usage?.let { usage.arrival.dayOfMonth })
        assertEquals(2, usage?.let { usage.arrival.monthValue })
        assertEquals(7, usage?.let { usage.departure.dayOfMonth })
        assertEquals(2, usage?.let { usage.departure.monthValue })
        assertEquals(7L, usage?.let { usage.days })
    }

    @Test
    fun deleteVisaUsageById_cascadeDeletesUsageFromUser() {
        userService.addVisaUsage("testuser@example.com", VisaUsageAddRequest("01/02/2020", "07/02/2020"))
        val updatedUser: User? = userService.getCurrentUser("testuser@example.com")
        assertEquals(1, updatedUser?.let { updatedUser.visaUsages.size })
        val usage: Long? = updatedUser?.let{updatedUser.visaUsages.first().id}
        userService.deleteVisaUsageById(usage)
        val userWithDeletedUsage: User? = userService.getCurrentUser("testuser@example.com")
        Assert.assertTrue(userWithDeletedUsage?.let { userWithDeletedUsage.visaUsages.isEmpty() } ?: false)
    }
}